### DISCUSSION TITLE
Is it possible to multiply to matricies and still get take the jacobian?

### DISCUSSION THREAD

**Discussion by Undermine28** [brianguenter/FastDifferentiation.jl#70] [Q&A] [answered]
I need to calculate the jacobian of the product of two matrices. When I multiply them, they become an array of Any and are no longer a FastDifferentiation.Node. Is this a limitation, or am I doing something wrong? The minimal example is as follows:

```
@variables a1 a2 b1 b2 c1 c2 d1 d2
thing1 = [a1 b1; c1 d1]
thing2 = [a2 b2; c2 d2]

thing3 = thing1 * thing2
jacobian(thing3, [a1, c2])
```

This fails with the error: 
> MethodError: no method matching jacobian(::Matrix{Any}, ::Vector{FastDifferentiation.Node})
> 
> Closest candidates are:
>   jacobian(!Matched::AbstractVector{T}, ::AbstractVector{S}) where {T<:FastDifferentiation.Node, S<:FastDifferentiation.Node}

Any help would be greatly appreciated

**Comment by brianguenter** [comment, marked as answer]
Yeah, this is a known issue. The compiler can't infer that Matrix{Node} * Matrix{Node} => Matrix{Node} even though I have defined all the arithmetic functions for Node, e.g., Node*Node => Node. Maybe there is a simple promotion rule that would make this work but I haven't figured it out yet. 

For now you can do this:
```julia
@variables a1 a2 b1 b2 c1 c2 d1 d2
thing1 = [a1 b1; c1 d1]
thing2 = [a2 b2; c2 d2]

thing3 = FastDifferentiation.Node.(thing1 * thing2) #convert from any to Node
jacobian(vec(thing3), [a1, c2]) #need to convert thing3 from matrix to vector since jacobian only supports vectors for now.
```

**Comment by brianguenter** [comment]
I think I've figured out how to fix this. I'll work on it over the next few days and if it is successful I'll make a patch release.

**Reply by Undermine28** [reply]
Ah of course! Thank you very much. I tried to call Node() which didn't work. I'm pretty new to Julia and did not yet understand the broadcast . Operator.

It makes sense now to call Node.()

**Comment by brianguenter** [comment]
I fixed this problem. In 20 minutes or so there should be a new patch release of FastDifferentiation on the Julia Registry, v0.3.8. If you update your packages this should fix the problem for you so you won't have to use the workaround of doing Node.() on your answer.

**Comment by Undermine28** [comment]
Thats great! Thanks for the quick resolution, Brain; much appreciated.
